server/src/test/java/org/elasticsearch/versioning/ConcurrentSeqNoVersioningIT.java

```diff
@@ -0,0 +1,905 @@
+    // back that a CAS failed due to version conflict even though it actually succeeded.
+    // 2. If we end up reporting back any other failure, it is unknown if the write succeeded or not (or will succeed in the future).
+    // 3. If you read data out, you may see dirty writes, ie. writes that will end up being discarded due to node failure/network
+    // disruption.
+    // 4. Likewise, the CAS check can be done against a dirty write and thus fail even if it ought to succeed.
+    // 5. A CAS check can be done against a stale primary and fail due to that.
+    // 6. A CAS failure does not give back the current seqno/primary-term to use for the next request (and if it did, it could be wrong
+    // due to a dirty write anyway).
+    public void testSeqNoCASLinearizability() {
+        // 1-8 seconds, bias towards shorter.
+        final int disruptTimeSeconds = 1 << randomInt(3);
+
+        assertAcked(prepareCreate("test")
+            .setSettings(Settings.builder()
+                .put(IndexMetaData.SETTING_NUMBER_OF_SHARDS, 1 + randomInt(2))
+                .put(IndexMetaData.SETTING_NUMBER_OF_REPLICAS, randomInt(3))
+            ));
+
+        ensureGreen();
+
+
+        int numberOfKeys = randomIntBetween(1,10);
+
+        logger.info("--> Indexing initial doc for {} keys", numberOfKeys);
+        List<Partition> partitions =
+            IntStream.range(0, numberOfKeys)
+                .mapToObj(i -> client().prepareIndex("test", "type", "ID:" + i).setSource("value", -1).get())
+                .map(response ->
+                    new Partition(response.getId(), new Version(response.getPrimaryTerm(), response.getSeqNo())))
+                .collect(Collectors.toList());
+
+        int threadCount = randomIntBetween(3, 20);
+        CyclicBarrier roundBarrier = new CyclicBarrier(threadCount + 1); // +1 for main thread.
+
+        List<CASUpdateThread> threads =
+            IntStream.range(0, threadCount)
+                .mapToObj(i -> new CASUpdateThread(i, roundBarrier, partitions, disruptTimeSeconds+1))
+                .collect(Collectors.toList());
+
+        logger.info("--> Starting {} threads", threadCount);
+        threads.forEach(Thread::start);
+
+        try {
+            int rounds = randomIntBetween(2, 5);
+
+            logger.info("--> Running {} rounds", rounds);
+
+            for (int i = 0; i < rounds; ++i) {
+                ServiceDisruptionScheme disruptionScheme = addRandomDisruptionScheme();
+                roundBarrier.await(1, TimeUnit.MINUTES);
+                disruptionScheme.startDisrupting();
+                logger.info("--> round {}", i);
+                try {
+                    roundBarrier.await(disruptTimeSeconds, TimeUnit.SECONDS);
+                } catch (TimeoutException e) {
+                    roundBarrier.reset();
+                }
+                internalCluster().clearDisruptionScheme(false);
```

Comment:
I chose to only ensureFullyConnected over also waiting for healthy since that part seems unnecessary here. First of all, it waits for healthy before doing `ensureFullyConnected`, so I guess that could delay test a bit. Second, I think the wait here is unnecessary and I prefer not to have it since the test should continue to run regardless of how many nodes master thinks are in the cluster and whethere there are any relocating shards.
In summary: next round could start earlier without the wait, which seems like it could provoke more failures.

Reply:
ok